proposals/kubeflow-conformance-program-proposal.md

```diff
@@ -0,0 +1,111 @@
+The [Kubeflow Brand Guidelines](https://github.com/kubeflow/community/blob/master/KUBEFLOW_BRAND_GUIDELINES.pdf) was published in Mar. 2021. The guideline is broadly applicable to usage of Kubeflow trademark by products, events and publications. While the brand guidelines provide general guidance, it does not prescribe the definition of Kubeflow and what makes a distribution/application "Kubeflow" vs. not "Kubeflow".  
+This document aims to define conformance criteria for the following usage of Kubeflow trademark:
+
+-  For Kubeflow distribution - "Certified Kubeflow"
+-  For Kubeflow application - TBD
+
+The goal is to ensure these special usages of Kubeflow trademark meet common standards of interoperability, increase cohesiveness of the Kubeflow platform, promote customer confidence, reduce the burden of Kubeflow maintainers, and extend Kubeflow’s influence beyond the Kubeflow project.
+
+# Kubeflow Distribution
+
+A conformant Kubeflow Distribution is certified to provide a set of core functionalities and API integration options. A core set of these features are listed in the "[Pipeline tests](#heading=h.71p7gyy4tf5d)" and "[Metadata tests](#heading=h.wc7nytwk964f)'' sections.  
+At the high level, the tests certifies that Kubeflow Pipeline and Metadata are part of the distribution, and exposes a standard set of APIs. Pipeline and Metadata and the binding "glue" for other Kubeflow components. With open standards for certifying Kubeflow Applications (see next section), the Kubeflow conformance criteria gives more flexibility to Kubeflow distributors to provide customization.   
+The tests will be designed in a way similar to [Kubernetes conformance program](https://github.com/cncf/k8s-conformance).  
+The tests will be implemented in stages.  
+Conformant distributions is entitled to refer to the distribution as "Certified Kubeflow". The distribution can be listed under a partner page under the Kubeflow project. The naming of the distribution still needs to follow [Kubeflow Brand Guidelines](https://github.com/kubeflow/community/blob/master/KUBEFLOW_BRAND_GUIDELINES.pdf).
+
+## Example
+
+Company X creates a distribution of Kubeflow and plans to name it "X Kubeflow Service". Company X tries to certify for Kubeflow conformance, which entails:
+
+-  Install the distribution
+-  Runs the Kubflow conformance test suite
+-  Submits the test log as a PR to Kubeflow Trademark Team for approval.
+-  Upon approval and following trademark guidelines, Company X changes the distribution name to "X service for Kubeflow", or "X Platform (Certified Kubeflow)"
+-  Optionally, the distribution may be listed in a catalog on Kubeflow website
+
+The conformance tests make sure "X service for Kubeflow" supports KFP and Metadata. Company X may include more applications (e.g. TFJob, Katib) in the distribution, but it does not affect the conformance standing of "X service for Kubeflow".
```

Comment:
Let's continue our discussion on the Kubeflow Community Meeting.

Why including Training Operators and Katib in the distribution conformance test is optional ?
Since these components are in Kubeflow core, does it make sense to verify that distribution is working with the core functionality and the APIs ?

From my point of view, Kubeflow distribution should support all Kubeflow core functionality to provide complete Cloud Native MLOps experience for the users.
In addition to that, it helps contributions to get a feedback from the distribution owners.

What do others think @kubeflow/wg-training-leads @kubeflow/wg-automl-leads @jbottum ?

Reply:
As you know, some platforms adopt a subset of Kubeflow components, should we have different conformance tests for different components?

Reply:
@andreyvelich I agree with you. Only including pipelines and metadata in the KF certification will confuse both developers and users. We specifically list multiple components as core components in KF (https://www.kubeflow.org/docs/components/) and just certifying one will cause confusion with no clear direction. @james-jwu are we only certifying pipelines because we have plans for other components similar to KFServing graduating?  It feels like this community is not transparent in a lot of strategic KF direction.

Reply:
One of the premises of KF Distribution Certification is that Kubeflow endeavored to be neutral and not endorse one distribution over another.  In addition, Kubeflow Distributions allows customization by distro vendors and provides the flexibility of choosing components to include in the distro. As Ce said, not all platforms adopt the full set of components currently under the Kubeflow project. With the introduction of Kubeflow Application Certification (proposed name), we envision many projects outside of Kubeflow project contributing the the Kubeflow ecosystem. KFServing’s graduation is one example where the certification program can easily accommodate projects not directly under Kubeflow.

The point about basing the tests on Pipeline and Metadata is not that they are the totality of “Kubeflow”.  Pipeline and Metadata provides the API and common layer that binds all the components. [quote from PR] “_Pipeline and Metadata are the binding "glue" for other Kubeflow components. With open standards for certifying Kubeflow Applications (see next section), the Kubeflow conformance criteria gives more flexibility to Kubeflow distributors to provide customization._” I also want to point to the long [discussion](https://groups.google.com/g/kubeflow-discuss/c/d6whgEgror8) we had last year, which raised some excellent points about why pipeline and metadata could be good choices for the minimal common layer that brings cohesion to Kubeflow ([[1]](https://groups.google.com/g/kubeflow-discuss/c/d6whgEgror8/m/yw1lHf5nBgAJ), [[2]](https://groups.google.com/g/kubeflow-discuss/c/d6whgEgror8/m/U9MYsyRTBgAJ)).
The point about providing a complete Cloud Native MLOps experience is well taken. I think most distro vendors will likely include most of the components under https://www.kubeflow.org/docs/components/ anyway. Forcing all distros to include all components is IMO not necessarily going to promote the growth of Kubeflow.

Reply:
> As Ce said, not all platforms adopt the full set of components currently under the Kubeflow project.

That's true, but one of the biggest advantages of using Kubeflow is that our users can deploy necessary components or deploy Kubeflow as an end-to-end MLOps platform. From the recent [Kubeflow User Panel](https://youtu.be/nFXL48x0Lk0), we saw that users want to have both options, right @jbottum ?
Conformance Program for each Kubeflow Core component might help to make distribution compatible with the Kubeflow Core.

As I can see from the thread discussion, the problem is that other Kubeflow components are loosely coupled.
Our future work should be focused on how to improve it, isn't it ?
For example, we are building integration between Katib and distributive Training Operators.

As well from the thread, I can point this message:
```
However, I imagine that the value of Kubeflow is offering a solution to the MLOps and I'd perhaps suggest even to the ResearchOps side of things, both of which are not a solved problem just yet.
```

Can Pipelines and Metadata offer end to end MLOps and ResearchOps infrastructure without other Kubeflow components ?

>  I think most distro vendors will likely include most of the components under https://www.kubeflow.org/docs/components/ anyway. 

Are we planing to add all components which pass Kubeflow Application Certification under https://www.kubeflow.org/docs/components ?

Reply:
@james-jwu I think I have a better understanding of why only pipelines and metadata are only included in the conformance test. You are basically saying to get the certified KF logo at a minimum get pipelines and metadata to pass conformance versus asking for all components to get certified. I don't agree with your statement "_Pipeline and Metadata are the binding "glue" for other Kubeflow components_" we have Data Scientists users that perceive Jupyter Notebooks as the "glue". From a devops/mlops perspective yes Pipelines are the "glue".  I have seen other communities offer levels or tiers of certification or conformance, for example Kubeflow conformance would be pipelines and metadata and Kubeflow+ would pipelines, metadata and all other components. I am not clear on what Kubeflow Application conformance will cover. Would it be possible to have a two tier conformance testing? I am concerned that excluding some Kubeflow components from conformance testing will slow adoption and cause user and developer confusion.

Reply:
Andrey, Juana, thanks for your feedback. You have some great points, and I think we can address them in an iterative way. One thing we learned from the Kubernetes Conformance program is to start small. The initial Kubernetes conformance tests ([v1.10](https://github.com/cncf/k8s-conformance/blob/master/docs/KubeConformance-1.10.md)) had only 174 tests, and [later versions](https://github.com/cncf/k8s-conformance) expanded that set significantly. We think Kubeflow Conformance will also evolve over time, and adapt to the needs of Kubeflow community, distributors and users.

I think starting the conformance program with the base interop layer (i.e. pipeline + metadata) sets a good path to an inclusive Kubeflow ecosystem, where distributors have assurance of the interoperability of components/applications that are bundled into the distro. We will probably learn a lot in the process. The feedback and information gathered from this effort can guide us to the next steps, which could expand the conformance scope.

With the initial scope of the conformance program, I think it is totally okay to produce a Certified distro that in its description has something along the lines of “Supports MLOps components under the official Kubeflow project: [https://www.kubeflow.org/docs/components](https://www.kubeflow.org/docs/components).”

The idea about 2 tiers of conformance is interesting. I agree that Notebook is an essential tool for data scientists. Tests for Notebook seem non-specific to Kubeflow, so the right scope of the tests is a little unclear to me at this point. What constitutes “core components” may warrant a discussion as well in the context of conformance.

RE - “Are we planning to add all components which pass Kubeflow Application Certification under https://www.kubeflow.org/docs/components” - the answer is no. Certified Application will have their own catalog, not to be confused with components living in the Kubeflow project.

Reply:
@james-jwu thank you very much for taking our input into consideration. I think it makes sense to start small with Pipelines and Metadata and plan for more components to be included later. As long as we are clear to our Users and Developers that more components testing will be included in the future and its part of the roadmap. Is it possible to add a sentence or section in the document reflecting the plan to include more KF components to conformance testing in later stages? Thanks

Reply:
I have raised some very similar concerns in my comment here: https://github.com/kubeflow/community/pull/524#issuecomment-916568247

@andreyvelich @gaocegege @johnugeorge @nakfour @james-jwu

Reply:
@nakfour Sounds good. I will provide an update after collecting more feedback.

Reply:
I added a few point we discussed in the previous community meeting. PTAL. @nakfour @thesuperzapper @animeshsingh @jbottum . The summary is:

- Kubeflow distribution
  - Added statement about expanding the conformance scope in the future
  - Added information about conformance versioning and expiration
  - Added out-of-scope issues for conformance tests: product quality/supportability, distribution development/channel
- Kubeflow application
  - Added the requirement that Kubeflow Application must be Kubernetes Application through self-attestation on deployment mechanism.